codesigning/gcp-cloudfunc/main.py

```diff
@@ -0,0 +1,175 @@
+
+    logging.debug('Data POSTed: %s', data)
+
+    if data['action'] == 'dequeue':
+        with get_lock():
+            queuefile = bucket.blob('queue.json')
+            queue_dict = json.loads(queuefile.download_as_string())
+            try:
+                next_file_url = queue_dict['queue'].pop(0)
+            except IndexError:
+                # No items in the queue!
+                logging.info('No binaries are currently queued for signing')
+                return jsonify('No items in the queue'), 204
+
+            queuefile.upload_from_string(json.dumps(queue_dict))
+
+        data = {
+            'https-url': next_file_url,
+            'basename': os.path.basename(next_file_url),
+            'gs-uri': unquote(next_file_url.replace(
+                f'{GOOGLE_PREFIX}/', 'gs://')),
+        }
+        logging.info(f'Dequeued {next_file_url}')
+        return jsonify(data)
+
+    elif data['action'] == 'enqueue':
+        url = data['url']
+        logging.info('Attempting to enqueue url" %s', url)
+
+        if not url.endswith('.exe'):
+            logging.info("Rejecting URL because it doesn't end in .exe")
+            return jsonify('Invalid URL to sign'), 400
+
+        if not url.startswith(GOOGLE_PREFIX):
+            logging.info('Rejecting URL because it does not start with %s',
+                         GOOGLE_PREFIX)
+            return jsonify('Invalid host'), 400
+
+        if not url.startswith((
+                f'{GOOGLE_PREFIX}/releases.naturalcapitalproject.org/',
+                f'{GOOGLE_PREFIX}/natcap-dev-build-artifacts/')):
+            logging.info('Rejecting URL because the bucket is incorrect')
+            return jsonify("Invalid target bucket"), 400
+
+        # Remove http character quoting
+        url = unquote(url)
+
+        binary_bucket_name, *binary_obj_paths = url.replace(
+            GOOGLE_PREFIX + '/', '').split('/')
+        codesign_bucket = storage_client.bucket(CODESIGN_DATA_BUCKET)
+
+        # If the file does not exist at this URL, reject it.
+        response = requests.head(url)
+        if response.status_code > 400:
+            logging.info('Rejecting URL because it does not exist')
+            return jsonify('Requested file does not exist'), 403
+
+        # If the file is too old, reject it.  Trying to avoid a
+        # denial-of-service by invoking the service with very old files.
+        mday, mmonth, myear = response.headers['Last-Modified'].split(' ')[1:4]
+        modified_time = datetime.datetime.strptime(
+            ' '.join((mday, mmonth, myear)), '%d %b %Y')
+        if modified_time < datetime.datetime(year=2024, month=6, day=1):
+            logging.info('Rejecting URL because it is too old')
+            return jsonify('File is too old'), 400
+
+        with get_lock():
+            # first, check to see if the file has already been signed.
+            signed_files_list = codesign_bucket.blob('signed_files.json')
```

Comment:
Does this track all files that have ever been signed?

Reply:
No, just the ones that have been signed by the service.  This seemed OK since if we request to sign an already-signed file, signing should just overwrite the existing signature with the new one.  In the future, it would be a nice enhancement to use `osslsigncode` to check and see if a file has already been signed and then add it to this file if so.

Reply:
Ah okay! If this is only used for releases, and not signing regular builds on main, this file shouldn't get prohibitively large